ingesters/AzureEventHubs/main.go

```diff
@@ -0,0 +1,423 @@
+			callback := func(ctx context.Context, msg *eventhubs.Event) error {
+				ent := &entry.Entry{
+					Data: msg.Data,
+					Tag:  tagid,
+					SRC:  src,
+				}
+				size += uint64(len(msg.Data))
+				if hubDef.Parse_Time == false {
+					if msg.SystemProperties != nil && msg.SystemProperties.EnqueuedTime != nil {
+						ent.TS = entry.FromStandard(*msg.SystemProperties.EnqueuedTime)
+					} else {
+						ent.TS = entry.Now()
+					}
+				} else {
+					ts, ok, err := tg.Extract(msg.Data)
+					if !ok || err != nil {
+						//  failed to extract, use the publishtime
+						hubDef.Parse_Time = false
+						if msg.SystemProperties != nil && msg.SystemProperties.EnqueuedTime != nil {
+							ent.TS = entry.FromStandard(*msg.SystemProperties.EnqueuedTime)
+						} else {
+							ent.TS = entry.Now()
+						}
+					} else {
+						ent.TS = entry.FromStandard(ts)
+					}
+				}
+				if err := procset.Process(ent); err != nil {
+					lg.Error("failed to process entry", log.KVErr(err))
+				}
+				count++
+				return nil
+			}
+
+			// get info about partitions in the hub
+			info, err := hub.GetRuntimeInformation(ctx)
+			if err != nil {
+				lg.Fatal("failed to get runtime info", log.KVErr(err))
+			}
+
+			// Launch a listener for each partition in the hub
+			// Calling Receive takes a while, but we can't really parallelize it because the first thing
+			// Receive does is lock a mutex in the Hub -- one way or another, it's basically serial.
+			for _, partitionID := range info.PartitionIDs {
+				// ask where to start from
+				checkpoint, err := diskPersist.Read(hubDef.Event_Hubs_Namespace, hubDef.Event_Hub, hubDef.Consumer_Group, partitionID)
+				if err != nil {
+					// set a default, we will check user setting next
+					checkpoint = persist.NewCheckpointFromStartOfStream()
+				}
+				if checkpoint.Offset == persist.StartOfStream && hubDef.Initial_Checkpoint == "end" {
+					checkpoint = persist.NewCheckpointFromEndOfStream()
+				}
+				handle, err := hub.Receive(
+					ctx,
+					partitionID,
+					callback,
+					eventhubs.ReceiveWithStartingOffset(checkpoint.Offset),
+				)
+				if err != nil {
+					lg.Error("failed to start event hub partition receiver", log.KVErr(err))
+					return
+				}
+				listeners = append(listeners, handle)
+				readers = append(readers, readerInfo{hubDef.Event_Hubs_Namespace, hubDef.Event_Hub, hubDef.Consumer_Group, partitionID})
+				lg.Info("started receiver for partition", log.KV("partition", partitionID))
+			}
+			<-quitSig
+		}(k, *def)
+	}
+
+	lg.Info("all hubs successfully started")
+
+	//register quit signals so we can die gracefully
+	utils.WaitForQuit()
+
+	lg.Info("exiting")
+	// Tell every event handler to close
+	for _, h := range listeners {
+		cctx, _ := context.WithTimeout(ctx, 2*time.Second)
+		h.Close(cctx)
+	}
+
+	// Tell our goroutines to bail out
+	close(quitSig)
+
+	// Write out persistence info one last time by hand.
```

Comment:
looks like there is a race here.  The "ticker/persister" routine could be writing to the disk at the same time as you.

I would reccomend an "exit" channel on that routine so that when you hit here you can wait for that routine to fully exit before performing the final write back to the disk.

Reply:
maybe not, the mem and disk persister have mutexes internally.